Fixes #13772 - orchestration now uses app logger

lzap opened the pull request:
Yeah, as simple as that. No more hidden orch issues in the flooded sql logger
which is turned off by default and in production environments it is almost
always turned off (while orch logging still provides a lot of very useful
messages at INFO/WARN level).

![image](https://cloud.githubusercontent.com/assets/49752/17623884/a1c688ce-60a2-11e6-82ac-6f7f3d6e9178.png)

lzap committed:
Fixes #13772 - orchestration now uses app logger

iNecas commented on app/models/concerns/orchestration/common.rb:
```diff
@@ -5,4 +5,9 @@ def handle_validation_errors
   end
+
+  # override sql logger for all orchestration mix-ins
+  def logger
```
Note that `Orchestration::Common` is included only to `DHCP` and `DNS` module. It it actually was included to all models that use orchestration (such as Host::Managed in https://github.com/theforeman/foreman/blob/develop/app/models/host/managed.rb#L53) would use the app logger, while others would use the `sql`. 

Maybe what we need is just to override the logger in all our models to use the `app` logger: it seems ActiveRecord doesn't use this method internally.

lzap replied:
> Maybe what we need is just to override the logger in all our models to use the app logger: it seems ActiveRecord doesn't use this method internally.

Do you have a trick in mind to do that?

lzap replied:
Yeah you are right, AR uses the new Rails 4 thing (notifications). I think I should be able to iterate through all our model classes and override the method. Any better ideas?

iNecas replied:
Katello uses a common class for all models of the plugin. https://github.com/Katello/katello/blob/master/app/models/katello/model.rb. Maybe time for Foreman to follow the same approach to influence the common logic. I assume you don't want to iterate thought the models dynamically, but rather putting the `include` statement there, right?

lzap commented:
Incorporated Ivan's idea. I can't use `Foreman::Model` because that's currently taken by the (Hardware) Model class, so going for `Foreman::BaseModel`. We also had an empty file `lib/foreman/model.rb` for some reason, removed.

domcleal commented:
Note that the Rails standard is `ApplicationRecord` in apps/models/application_record.rb (https://github.com/rails/rails/blob/master/railties/lib/rails/generators/rails/app/templates/app/models/application_record.rb).

dLobatog commented:
Maybe it's best to add that `logger` method to ApplicationRecord..?

lzap commented:
Good idea, done.

iNecas commented:
Not tested but conceptually :+1:, this will help A LOT! Thanks @lzap

ares commented on lib/foreman/model.rb:
```diff
@@ -1,2 +0,0 @@
-module Foreman::Model
-end
```
I think this is still being used at [compute resources](https://github.com/theforeman/foreman/blob/develop/app/models/compute_resources/foreman/model/vmware.rb#L4), we should probably keep it or change the definition there to

``` ruby
module Foreman
  module Mode
  ...
```

lzap replied:
Hmm this would change indentation of many files, I will put it back with comment:

```
app/models/compute_resources/foreman/model/ovirt.rb
4:module Foreman::Model

app/models/compute_resources/foreman/model/openstack.rb
1:module Foreman::Model

app/models/compute_resources/foreman/model/gce.rb
1:module Foreman::Model

app/models/compute_resources/foreman/model/vmware.rb
4:module Foreman::Model

app/models/compute_resources/foreman/model/ec2.rb
1:module Foreman::Model

app/models/compute_resources/foreman/model/libvirt.rb
1:module Foreman::Model

app/models/compute_resources/foreman/model/rackspace.rb
1:module Foreman::Model
```

ares replied:
it seems it's still being removed

ares commented:
LGTM except one nit, did you think about some way to let plugins know that they should start using the new abstract class? Maybe just documenting at [wiki](http://projects.theforeman.org/projects/foreman/wiki/How_to_Create_a_Plugin) and foreman-dev mailing list announcement would be enough

lzap commented:
Updated, updating the wiki as well. Rebased.

lzap commented:
http://projects.theforeman.org/projects/foreman/wiki/How_to_Create_a_Plugin#Adding-new-model-classes

ares commented:
it breaks many tests, maybe abstract_class is missing in some test models?

lzap commented:
I rebased, anyone knows why this fails on Jenkins? Works fine locally.

iNecas commented:
Any chance to revive this one?

lzap commented:
I rebased, @ares ?

ares commented:
[test foreman]

Katello tests are failing, seems related
```
14:16:12 == 20140610083129 AddPulpProxyToHost: migrating ===============================
14:16:12 -- add_column(:hosts, :content_source_id, :integer)
14:16:12    -> 0.0008s
14:16:12 -- add_column(:hostgroups, :content_source_id, :integer)
14:16:12    -> 0.0004s
14:16:12 -- add_index(:hosts, :content_source_id)
14:16:12    -> 0.0032s
14:16:12 -- add_index(:hostgroups, :content_source_id)
14:16:12    -> 0.0027s
14:16:12 -- add_foreign_key(:hosts, :smart_proxies, {:name=>"hosts_content_source_id_fk", :column=>"content_source_id"})
14:16:12    -> 0.0014s
14:16:12 -- add_foreign_key(:hostgroups, :smart_proxies, {:name=>"hostgroups_content_source_id_fk", :column=>"content_source_id"})
14:16:12    -> 0.0012s
14:16:12 == 20140610083129 AddPulpProxyToHost: migrated (0.0101s) ======================
14:16:12 
14:16:12 rake aborted!
14:16:12 TypeError: superclass mismatch for class Environment
14:16:12 /var/lib/workspace/workspace/test_develop_pr_katello/database/postgresql/ruby/2.2/slave/fast/plugin/db/migrate/20140610154745_content_view_puppet_environment_id.rb:2:in `<class:ContentViewPuppetEnvironmentId>'
14:16:12 /var/lib/workspace/workspace/test_develop_pr_katello/database/postgresql/ruby/2.2/slave/fast/plugin/db/migrate/20140610154745_content_view_puppet_environment_id.rb:1:in `<top (required)>'
14:16:12 /usr/local/rvm/gems/ruby-2.2.5@test_develop_pr_katello-1/gems/polyglot-0.3.5/lib/polyglot.rb:65:in `require'
14:16:12 /usr/local/rvm/gems/ruby-2.2.5@test_develop_pr_katello-1/gems/polyglot-0.3.5/lib/polyglot.rb:65:in `require'
14:16:12 /usr/local/rvm/gems/ruby-2.2.5@test_develop_pr_katello-1/gems/activesupport-4.2.8/lib/active_support/dependencies.rb:274:in `block in require'
14:16:12 /usr/local/rvm/gems/ruby-2.2.5@test_develop_pr_katello-1/gems/activesupport-4.2.8/lib/active_support/dependencies.rb:240:in `load_dependency'
14:16:12 /usr/local/rvm/gems/ruby-2.2.5@test_develop_pr_katello-1/gems/activesupport-4.2.8/lib/active_support/dependencies.rb:274:in `require'
14:16:12 /usr/local/rvm/gems/ruby-2.2.5@test_develop_pr_katello-1/gems/activerecord-4.2.8/lib/active_record/migration.rb:777:in `load_migration'
14:16:12 /usr/local/rvm/gems/ruby-2.2.5@test_develop_pr_katello-1/gems/activerecord-4.2.8/lib/active_record/migration.rb:773:in `migration'
14:16:12 /usr/local/rvm/gems/ruby-2.2.5@test_develop_pr_katello-1/gems/activerecord-4.2.8/lib/active_record/migration.rb:768:in `disable_ddl_transaction'
14:16:12 /usr/local/rvm/gems/ruby-2.2.5@test_develop_pr_katello-1/gems/activerecord-4.2.8/lib/active_record/migration.rb:1051:in `use_transaction?'
14:16:12 /usr/local/rvm/gems/ruby-2.2.5@test_develop_pr_katello-1/gems/activerecord-4.2.8/lib/active_record/migration.rb:961:in `rescue in block in migrate'
14:16:12 /usr/local/rvm/gems/ruby-2.2.5@test_develop_pr_katello-1/gems/activerecord-4.2.8/lib/active_record/migration.rb:958:in `block in migrate'
14:16:12 /usr/local/rvm/gems/ruby-2.2.5@test_develop_pr_katello-1/gems/activerecord-4.2.8/lib/active_record/migration.rb:955:in `each'
14:16:12 /usr/local/rvm/gems/ruby-2.2.5@test_develop_pr_katello-1/gems/activerecord-4.2.8/lib/active_record/migration.rb:955:in `migrate'
14:16:12 /usr/local/rvm/gems/ruby-2.2.5@test_develop_pr_katello-1/gems/activerecord-4.2.8/lib/active_record/migration.rb:823:in `up'
14:16:12 /usr/local/rvm/gems/ruby-2.2.5@test_develop_pr_katello-1/gems/activerecord-4.2.8/lib/active_record/migration.rb:801:in `migrate'
14:16:12 /usr/local/rvm/gems/ruby-2.2.5@test_develop_pr_katello-1/gems/activerecord-4.2.8/lib/active_record/tasks/database_tasks.rb:137:in `migrate'
14:16:12 /usr/local/rvm/gems/ruby-2.2.5@test_develop_pr_katello-1/gems/activerecord-4.2.8/lib/active_record/railties/databases.rake:44:in `block (2 levels) in <top (required)>'
14:16:12 /usr/local/rvm/gems/ruby-2.2.5@test_develop_pr_katello-1/gems/rake-12.0.0/exe/rake:27:in `<top (required)>'
14:16:12 /usr/local/rvm/gems/ruby-2.2.5@test_develop_pr_katello-1/bin/ruby_executable_hooks:15:in `eval'
14:16:12 /usr/local/rvm/gems/ruby-2.2.5@test_develop_pr_katello-1/bin/ruby_executable_hooks:15:in `<main>'
14:16:12 TypeError: superclass mismatch for class Environment
```

lzap commented:
Rebased, I am filing issue/PR to fix Katello right now.

ares commented:
Foreman tests unrelated

ares commented:
Works nicely, ready for merge when related PRs are ready.

lzap commented:
Thanks I am going to review other plugins now, but not all occurances of `ActiveRecord::Base` means a failure. For example Discovery patch is not needed, it works fine as well as unit tests.

iNecas commented:
👍 @bbuckingham @ehelms as Katello seems to be likely affected, I would like to get your response on https://github.com/Katello/katello/issues/6700 before we merge, but merging this will greatly improve our ability to investigate provisioning issues

lzap commented:
Let's wait after branching for this one.

lzap commented:
Rebased, since we branched off, please merge. I suggest manual merge, rebase and grep for `ActiveRecord::Base` in the codebase for *relevant* occurances (if new model was added for example). There are few irrelevant items.

lzap commented:
Guys I need some cooperation here, I've been rebasing this several times already, Katello tests will not pass. I need someone to commit to merging this and then I will rebase, thanks.

lzap commented:
I have just verified this, does not need a rebase, good to go:

```
$ ag "ActiveRecord::Base"
app/controllers/api/v2/base_controller.rb
151:        ActiveRecord::Base.include_root_in_json = false
155:        ActiveRecord::Base.include_root_in_json = true

app/services/dashboard/data.rb
23:    con = ActiveRecord::Base.connection

app/services/fact_importer.rb
40:    ActiveRecord::Base.transaction do
65:      if ActiveRecord::Base.connection.adapter_name.downcase.starts_with? 'mysql'

app/validators/bookmark_controller_validator.rb
2:  @@active_record_tables = ActiveRecord::Base.connection.tables.map(&:to_s)

app/models/application_record.rb
1:class ApplicationRecord < ActiveRecord::Base

app/models/concerns/scoped_search_extensions.rb
17:      is_pg = ActiveRecord::Base.connection.adapter_name.downcase.starts_with? 'postgresql'

config/initializers/active_record_extensions.rb
1:class ActiveRecord::Base

lib/core_extensions.rb
9:class ActiveRecord::Base

lib/tasks/log.rake
3:  ActiveRecord::Base.logger = Logger.new(STDOUT)

lib/tasks/convert.rake
72:      class ProductionModelClass < ActiveRecord::Base
76:      class DevelopmentModelClass < ActiveRecord::Base
81:      ActiveRecord::Base.establish_connection(:production)
83:      (ActiveRecord::Base.connection.tables - skip_tables).each do |table_name|

db/migrate/20130520161514_add_unique_constraint_to_fact_name.rb
4:    options = if ActiveRecord::Base.connection.instance_values["config"][:adapter].grep(/mysql/).any?

db/migrate/20130228145456_add_digest_to_sources.rb
3:    if ["mysql", "mysql2"].include? ActiveRecord::Base.connection.instance_values["config"][:adapter]
16:    if ["mysql", "mysql2"].include? ActiveRecord::Base.connection.instance_values["config"][:adapter]

db/migrate/20130520152000_remove_duplicate_fact_names.rb
6:      ActiveRecord::Base.transaction do

db/migrate/20130121130826_add_digest_to_messages.rb
3:    if ActiveRecord::Base.connection.instance_values["config"][:adapter] == "mysql" || ActiveRecord::Base.connection.instance_values["config"][:adapter] == "mysql2"

db/migrate/20090915030726_change_report_field_type_to_text.rb
3:    if ActiveRecord::Base.connection.instance_values["config"][:adapter] == "mysql" || ActiveRecord::Base.connection.instance_values["config"][:adapter] == "mysql2"
9:    if ActiveRecord::Base.connection.instance_values["config"][:adapter] == "mysql" || ActiveRecord::Base.connection.instance_values["config"][:adapter] == "mysql2"

db/migrate/20100325142616_update_fact_names_and_values_to_bin.rb
3:    if ActiveRecord::Base.connection.instance_values["config"][:adapter] == "mysql" || ActiveRecord::Base.connection.instance_values["config"][:adapter] == "mysql2"
10:    if ActiveRecord::Base.connection.instance_values["config"][:adapter] == "mysql" || ActiveRecord::Base.connection.instance_values["config"][:adapter] == "mysql2"

db/migrate/20100115021803_change_mysql_reports_column.rb
3:    if ActiveRecord::Base.connection.instance_values["config"][:adapter] == "mysql" || ActiveRecord::Base.connection.instance_values["config"][:adapter] == "mysql2"
9:    if ActiveRecord::Base.connection.instance_values["config"][:adapter] == "mysql" || ActiveRecord::Base.connection.instance_values["config"][:adapter] == "mysql2"

db/migrate/20130908170524_add_keys.rb
4:    if ActiveRecord::Base.connection.adapter_name == 'PostgreSQL'
6:    elsif ActiveRecord::Base.connection.adapter_name.downcase.starts_with? 'mysql'
96:    if ActiveRecord::Base.connection.adapter_name.downcase.starts_with? 'mysql'

db/migrate/20110321070954_revert_face_names_and_values_to_text_records.rb
3:    if ActiveRecord::Base.connection.instance_values["config"][:adapter] == "mysql" || ActiveRecord::Base.connection.instance_values["config"][:adapter] == "mysql2"
9:    if ActiveRecord::Base.connection.instance_values["config"][:adapter] == "mysql" || ActiveRecord::Base.connection.instance_values["config"][:adapter] == "mysql2"

db/migrate/20101018120603_create_sources.rb
6:    if ActiveRecord::Base.connection.instance_values["config"][:adapter] == "mysql" || ActiveRecord::Base.connection.instance_values["config"][:adapter] == "mysql2"

db/migrate/20131021152315_change_name_index_on_fact_name.rb
4:    options = if ActiveRecord::Base.connection.instance_values["config"][:adapter].grep(/mysql/).any?
14:    options = if ActiveRecord::Base.connection.instance_values["config"][:adapter].grep(/mysql/).any?

db/migrate/20170118142654_add_auto_increment_to_bigint_ids.rb
3:    if ['mysql', 'mysql2'].include? ActiveRecord::Base.connection.instance_values['config'][:adapter]
10:    if ['mysql', 'mysql2'].include? ActiveRecord::Base.connection.instance_values['config'][:adapter]

db/migrate/20170110113824_change_id_value_range.rb
3:    if ['mysql', 'mysql2', 'postgresql'].include? ActiveRecord::Base.connection.instance_values['config'][:adapter]
10:    if ['mysql', 'mysql2', 'postgresql'].include? ActiveRecord::Base.connection.instance_values['config'][:adapter]

db/migrate/20101018120548_create_messages.rb
6:    if ActiveRecord::Base.connection.instance_values["config"][:adapter] == "mysql" || ActiveRecord::Base.connection.instance_values["config"][:adapter] == "mysql2"

test/controllers/api/v2/domains_controller_test.rb
33:    skip if ActiveRecord::Base.connection_config[:adapter].eql?"sqlite3"

test/controllers/hosts_controller_test.rb
1244:    ActiveRecord::Base.any_instance.expects(:destroy).never
1245:    ActiveRecord::Base.any_instance.expects(:save).never
1255:    ActiveRecord::Base.any_instance.expects(:destroy).never
1256:    ActiveRecord::Base.any_instance.expects(:save).never
1266:    ActiveRecord::Base.any_instance.expects(:destroy).never
1267:    ActiveRecord::Base.any_instance.expects(:save).never
1277:    ActiveRecord::Base.any_instance.expects(:destroy).never
1278:    ActiveRecord::Base.any_instance.expects(:save).never

test/unit/ensure_no_cycle_test.rb
11:    @graph = ActiveRecord::Base::EnsureNoCycle.new(base, :source, :target)

test/unit/tasks/seeds_test.rb
121:    ActiveRecord::Base.any_instance.expects(:save).never

test/facet_test_helper.rb
1:unless ActiveRecord::Base.connection.table_exists?(:test_facets)
2:  ActiveRecord::Base.connection.create_table :test_facets do |t|
7:unless ActiveRecord::Base.connection.table_exists?(:module_test_facets)
8:  ActiveRecord::Base.connection.create_table :module_test_facets do |t|

test/models/hostgroup_test.rb
440:      ActiveRecord::Base.any_instance.expects(:destroy).never
441:      ActiveRecord::Base.any_instance.expects(:save).never

test/models/host_test.rb
2639:      ActiveRecord::Base.any_instance.expects(:destroy).never
2640:      ActiveRecord::Base.any_instance.expects(:save).never

test/models/compute_resource_test.rb
211:    unless ActiveRecord::Base.connection.instance_values["config"][:adapter] == 'sqlite3'

test/test_helper.rb
31:if ActiveRecord::Base.connection.adapter_name == 'PostgreSQL'
```

iNecas commented:
[test foreman]

lzap commented:
It failed, not sure if this is related, rebased.

iNecas commented:
Thanks @lzap!
